### PR TITLE
Fix the shell error during agent version upgrade

### PR DESCRIPTION
**Pull request by yshapiro-57**
### What does this PR do?

This fixes a glitch introduced in https://github.com/DataDog/datadog-agent/pull/23084. That PR has since been rolled back, and I am reapplying it in this one in addition to fixing the error.

The glitch was that I copied the code [from the Linux install script](https://github.com/DataDog/agent-linux-install-script/blob/main/install_script.sh.template#L1418) without realizing that the Linux install script is written in `bash`, and the package install scripts are in `sh`, and there are syntax differences between the two. I addressed some differences in the original PR, but missed some, and that is causing errors when a user uses the package scripts to upgrade a version of the agent.

There are two syntax differences I am addressing in the forward fix:
- [Per this Stack Overflow answer](https://stackoverflow.com/a/10849346), strings should be compared with `=` instead of `==` in `sh`
- [Per this answer](https://stackoverflow.com/a/5917742), you cannot do this in `sh`:
```
contents=$(cat /tmp/file.txt)
substring=${contents:2:5}
```
It does not like extracting a substring from an expression that's not a name or a number. You have to instead use a different method like `awk` to extract a substring from a file.

I am also making the install script write `install.json` only if one does not already exist, because if one has already been written by the Linux agent install script, then we do not want to overwrite it.

### Motivation

This addresses #incident-25603 and is a part of the work on this Jira task: [Write an install signature during yum and apt-get installs](https://datadoghq.atlassian.net/browse/AIT-9866)

### Describe how to test/QA your changes

Create a new AWS EC2 Debian instance, and install the 7.51 build of the Datadog agent on it by running
```
DD_API_KEY=YOUR_PRODUCTION_API_KEY DD_SITE="datadoghq.com"  bash -c "$(curl -L https://s3.amazonaws.com/dd-agent/scripts/install_script_agent7.sh)"
```
Then rerun the test plan from https://github.com/DataDog/datadog-agent/pull/23084 on that instance, which causes the agent to get upgraded to 7.53. Without this fix you will get a "Bad substitution" shell error. With this fix there should be no error, and upgrading the agent should not change the install ID and install time in install.json.

Also verify that the original test plan from https://github.com/DataDog/datadog-agent/pull/23084 still passes.

To verify that the Linux agent installer still writes the correct install type with these changes, run the following commands on a brand new Ubuntu VM (this must be done within 24 hours of CI running on this PR):
```
export TESTING_APT_URL=apttesting.datad0g.com
export TESTING_APT_REPO_VERSION="pipeline-30133501-a7-x86_64 7"
DD_API_KEY=YOUR_PRODUCTION_API_KEY DD_SITE="datadoghq.com" bash -c "$(curl -L https://s3.amazonaws.com/dd-agent/scripts/install_script_agent7.sh)"
```
This will install the new version of the package that was built off the code in this PR. You will be able to confirm that `install.json` still shows the Linux install script and not the manual installer:
```
ubuntu@ip-10-112-103-253:~$ cat /etc/datadog-agent/install.json
{"install_id":"365ef5c1-fff2-4652-95c7-be297c93d792","install_type":"linux_agent_installer","install_time":1710451836}
ubuntu@ip-10-112-103-253:~$
```

### CODE REVIEW

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

**Comment by spencergilbert** [comment]
Because this telemetry data isn't required for the package, we're thinking we should ensure the script doesn't fail if there's an issue writing the file. There's already a section in the script that toggles the exit behaviour ([here](https://github.com/DataDog/datadog-agent/blob/1ff89cd23e4c5b135954eee659fc33586d4f0bcd/omnibus/package-scripts/agent-deb/postinst#L23-L26)) or include a `|| :` to ignore the error.

What do you think? I don't want to stop us from collecting this data, but we really don't want to cause package issues here.

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

File: omnibus/package-scripts/agent-deb/postinst [hunk 1]
```diff
@@ -65,6 +65,38 @@ install_method:
...
+    install_time="$3"
+    printf "{\"install_id\":\"$install_id\",\"install_type\":\"$install_type\",\"install_time\":$install_time}"
+}
+
+if [ ! -f "$CONFIG_DIR/install.json" ]; then
+    install_id=$(generate_install_id)
+    install_type="linux_manual_with_apt"
+    install_time=$(date +%s)
+    install_signature=$(generate_install_signature "$install_id" "$install_type" "$install_time")
+    sh -c "echo '$install_signature' > $CONFIG_DIR/install.json"
```
Comment (bkabrda) [inline omnibus/package-scripts/agent-deb/postinst:95]: Small question: Why are we using a subshell to execute this echo command?

Comment (yshapiro-57) [reply omnibus/package-scripts/agent-deb/postinst:95]: I think this was a copy and paste from the Linux agent install script. Let me remove the `sh -c`

Comment (spencergilbert) [reply omnibus/package-scripts/agent-deb/postinst:95]: Probably unnecessary to mention, but same applies for the two other postinst files

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Review by bkabrda — APPROVED** [review]

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Comment by yshapiro-57** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

**Pull request merged**